Jk/cumulus 3323 - Enforce SSL DB connections by default

Jkovarik opened the pull request:
**Summary:** This PR:

- Updates the default cluster creation to disallow non-ssl connections to the main database cluster. 
- Updates core's database package to:
   - Default to making an SSL connection and disallow unrecognized/self-signed certs
   - Allow configuration options `disableSSL`/`rejectUnauthorized` to allow Core to run without SSL and/or allow unrecognized CAs.
 - Addresses inconsistent integration test titles discovered while doing this work. 


Addresses [CUMULUS-3323](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3323)

## Changes

### Breaking Changes

- **CUMULUS-3323**
  - Updated `@cumulus/db` to by default set the `ssl` option for knex, and
    reject non-SSL connections via use of the `rejectUnauthorized` configuration
    flag.   This causes all Cumulus database connections to require SSL (CA or
    self-signed) and reject connectivity if the database does not provide SSL.
    Users using serverless v1/`cumulus-rds-tf` should not be impacted by this
    change as certs are provided by default.

### Changed

- **CUMULUS-3323**
  - Added `disableSSL` as a valid database secret key - setting this in your database credentials will
    disable SSL for all Core database connection attempts.
  - Added `rejectUnauthorized` as a valid database secret key - setting
    this to `false` in your database credentials will allow self-signed certs/certs with an unrecognized authority.
  - Updated the default parameter group for `cumulus-rds-tf` to set `force_ssl`
    to 1.   This setting for the Aurora Serverless v1 database disallows non-SSL
    connections to the database, and is intended to help enforce security
    compliance rules.  This update can be opted-out by supplying a non-default
    `db_parameters` set in the terraform configuration.

### Fixed

- **CUMULUS-3323**
  - Minor edits to errant integration test titles (dyanmo->postgres)

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

Jkovarik commented:
Tested this via manipulation of the DB login secrets/locally/via units.    AWS Serverless database was reconfigured to force SSL, validate that connections failed with SSL disabled.

npauzenga commented on lambdas/db-provision-user-database/tests/test-index.js:
```diff
@@ -15,6 +15,7 @@ test.before(async (t) => {
         username: localStackConnectionEnv.PG_USER,
         password: localStackConnectionEnv.PG_PASSWORD,
         database: localStackConnectionEnv.PG_DATABASE,
+        disableSSL: 'true',
```
So we're disabling ssl in tests because that's not explicitly what we want to test and presumably requires a cert and some setup?

Jkovarik replied:
Exactly that - we're not trying to test that knex connects via SSL properly, we're using the db as a stand in instead of mocking to test the logic.

npauzenga commented on packages/db/src/config.ts:
```diff
@@ -5,11 +5,12 @@ import { Knex } from 'knex';
-  PG_DATABASE: 'postgres',
   PG_PORT: '5432',
+  PG_USER: 'postgres',
+  DISABLE_PG_SSL: 'true',
```
What is the reasoning for having this default to disabled? I don't disagree, just want to understand.

npauzenga replied:
Oh because localstack and we don't want to mess with SSL config locally.

Jkovarik replied:
Exactly that - setting up SSL on a localstack postgres database is cumbersome.

npauzenga commented on packages/db/src/config.ts:
```diff
@@ -33,24 +34,39 @@ export const getSecretConnectionConfig = async (
       throw new Error(`AWS Secret ${SecretId} is missing required key '${key}'`);
```
Note for this function, not this line: could we add a docstring?

Jkovarik replied:
We can - I'm less inclined to in Typescript generally if the function is self-explanatory as it's self-annotated already (typing is duplicative). 

That said: ea8ae4fd6d6

npauzenga replied:
Hmm yeah maybe if the arguments are (properly) typed we don't need a docstring. Descriptions are handy sometimes but also prone to going stale. And in this case, obvious.

npauzenga commented on packages/db/src/config.ts:
```diff
@@ -33,24 +34,39 @@ export const getSecretConnectionConfig = async (
       throw new Error(`AWS Secret ${SecretId} is missing required key '${key}'`);
     }
   });
+  const rejectUnauthorized = dbAccessMeta.rejectUnauthorized !== 'false' && dbAccessMeta.rejectUnauthorized !== false;
```
So we don't know the type coming from the secret? 🤔 is it worth casting a string as a bool and enforcing types here rather than checking multiple options?

Jkovarik replied:
We can only infer typing from what we expect the user input to be.   Users taking a json bool and turning it into a string is entirely possible.      The typing coming back from JSON.parse is always `any` also, as it's user input essentially.  

We're *functionally* typing `rejectUnauthorized` here via an explicit check. 

Doing something like 

```
Boolean(dbAccessMeta.rejectUnauthorized) 
```

won't work because 

```
> Boolean("false")
true
```

I believe there's lodash options for truthy strings, but that felt like overhead here.

npauzenga replied:
All good. Checking string and bool is fine, if a little cumbersome, but that's what we have as user input. Thanks for thinking about it.

npauzenga commented on packages/db/src/config.ts:
```diff
@@ -33,24 +34,39 @@ export const getSecretConnectionConfig = async (
     }
   });
+  const rejectUnauthorized = dbAccessMeta.rejectUnauthorized !== 'false' && dbAccessMeta.rejectUnauthorized !== false;
+  const useSsl = dbAccessMeta.disableSSL === 'true' || dbAccessMeta.disableSSL === true;
```
Same as above. This is probably fine but is less cool than being able to assume a type 😄

Jkovarik replied:
See comment above - `useSsl` is assumed to be a bool by TS.

npauzenga commented on packages/db/src/config.ts:
```diff
@@ -33,24 +34,39 @@ export const getSecretConnectionConfig = async (
     password: dbAccessMeta.password,
-    database: dbAccessMeta.database,
     port: dbAccessMeta.port ?? 5432,
+    ssl: useSsl ? undefined : { rejectUnauthorized },
```
So `useSsl` is `true` if `dbAccessMeta.disableSSL === 'true'`? 🤯 

Then if `useSsl` isn't explicitly `true`, we default to `rejectUnauthorized`. That I think makes sense but the first part is causing an aneurysm.

Jkovarik replied:
Oof, that's .... just bad naming.  Try: 0f0e36e0219

npauzenga replied:
Thanks 😄

npauzenga commented on packages/db/src/config.ts:
```diff
@@ -33,24 +34,39 @@ export const getSecretConnectionConfig = async (
+): Knex.PgConnectionConfig => {
+  const rejectUnauthorized = env.REJECT_UNAUTHORIZED !== 'false';
+  const connectionConfigEnv: {
+    host: string,
```
👍 for typing

npauzenga commented on CHANGELOG.md:
```diff
@@ -14,19 +14,38 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
   - Removed unused CloudWatch AWS SDK client. This change removes the CloudWatch client
     from the `@cumulus/aws-client` package.
+- **CUMULUS-3323**
+  - Updated `@cumulus/db` to by default set the `ssl` option for knex, and
```
Do we need doc updates for this?

Jkovarik replied:
Hmm, yeah that's a good point.   How about : 2ebe3b1cff3

npauzenga replied:
Great, thanks! I think this covers it and doesn't get bogged down discussing why you'd use a self-signed cert, etc.

npauzenga reviewed (changes requested):
Looks great @Jkovarik. A couple clarifying questions but nothing major.

Jkovarik commented:
@npauzenga  I believe I've addressed your comments, thanks for the review!   Please take a look and let me know what you think.   If you'd like to pair on functional testing or would like a demo, reach out.
